### DISCUSSION TITLE
Using ASReview from R?

### DISCUSSION THREAD

**Discussion by rkrug** [asreview/asreview#2380] [Q&A]
Hi

I am an (very?) experienced R developer, and I would like to use ASReview from R. My usage scenario is that I am downloading larger datasets (up to several million)) from OpenAlex and would like to provide a reproducible (as much as possible) workflow to select relevant papers. I get, as needed, a list of DOIs / OpenAlexIDs which are relevant, others which are not relevant. I would now like to use ASReview to classify the others as relevant of not relevant, and in an additional step, draw a kind of 2d rarefaction surface indicating the relevant papers in relation to the included / excluded paper. 

I am completely stuck with python. I can create the project, but I am completely stuck in the second step. Here are the two files. I feel a bit lost at the moment. These were written by ChatGPT, so they might be for a python user horrendously complicated - but as I do not know any python, I wanted to use them as a first step.

As I said, the whole would be done from R. 

Any suggestions if this is the right approach? Cul de Sac?

Thanks for any input

I can provide the other files I use if somebody is interested. 

[03_rank_asreview_csv_v2.py](https://github.com/user-attachments/files/23342162/03_rank_asreview_csv_v2.py)
[simulate_rarefaction2d_from_project.py](https://github.com/user-attachments/files/23342163/simulate_rarefaction2d_from_project.py)
[config.yaml](https://github.com/user-attachments/files/23342215/config.yaml)

**Comment by Rensvandeschoot** [comment]
Great you are interested in using ASReview for simulation studies! Are you familiar with our infrastructure via ASReview-Makita ([scientific paper](https://doi.org/10.1016/j.simpa.2024.100663), [GitHub repo](https://github.com/asreview/asreview-makita)) and the [Insights ](https://github.com/asreview/asreview-insights)package providing you with all kinds of metrics? . The output files, if yu use the Makita workflow, will provide you with CSV files (for example the [time to discovery](https://systematicreviewsjournal.biomedcentral.com/articles/10.1186/s13643-023-02257-7) for each record) which you can further process in R. 

For a tutorial go to [ASReview-Academy](https://asreview.github.io/asreview-academy/simulation.html)

**Comment by rkrug** [comment]
Let me clarify my workflow - because I think that ASReview-Makita does not cover it fully (at the moment).

I essentially start with a partially labelled dataset in the range of million records.

The steps would be:

### Step A
1. create a ASReview project
2. vectorise all papers using probably sbert or another semantic one, as thes seems to be the most memory efficient ones, and cache the results
[03_rank_asreview_csv_v2.py](https://github.com/user-attachments/files/23355478/03_rank_asreview_csv_v2.py)



### Step B
1. Bootstrap the prior, i.e. use only a subset of the included and excluded labelled articles as priors
2. classify all papers automatically using the bootstrapped prior knowledge (and likely a threshold)
3. repeat this several times and for different numbers of included / excluded and get the number of automatically included paper, also which ones are included, to look at "stability" of the selection
[simulate_rarefaction2d_from_project.py](https://github.com/user-attachments/files/23355483/simulate_rarefaction2d_from_project.py)


### Step C
Use these as a kind of saturation curve to identify if more priors are needed for a stable selection

As far as I understand, simulations work on a fully labelled datasets? Otherwise they would be what I would need.

This is why I started drafting my own scripts in python (using ChatGPT).

I also use a config file ([asreview_config_template.yaml](https://github.com/user-attachments/files/23355501/asreview_config_template.yaml)) to keep everything completely reproducible (unless this info is stored in a ASReview project?), but this is not relevant at the moment.

**Reply by Rensvandeschoot** [reply]
So, it is a two step approach: 

1. using the labelled part train a model (you can use the web-app, the API or Makita)
2. then use it for starting a new project for unlabelled data

But, if you want to answer your questions in step C, you would need the labels for those records as well, right? Otherwise, you cannot know if you have identified all relevant records. And such a set-up is exactly one of the templates in the Makita workflow.

**Reply by rkrug** [reply]
OK - I think you are speaking of the `prior` template?

As I have only one dataset at the moment, I would have to split it into one labeled dataset `prior_dataset.csv` which contains the labelled records,and one dataset unlabelled `dataset.csv` which only contains the unlabelled records?

afterwards, I can use `asreview makita template prior` to get the scripts - correct? 

When I have the scripts, I can wrap them in R commands. If I execute them in the same order as the jobs.sh, I can replace the data with different datasets, and run it again?

As I have millions of records, it would be not feasible to re-vectorize the them for each simulation, so I would like to understand a little bit better what these are doing - I have adapted the script to my python installation and environment:

```
# version 1.1

# Create folder structure. By default, the folder 'output' is used to store output.
mkdir output
mkdir output/simulation
mkdir output/simulation/metrics
mkdir output/figures
mkdir output/tables/metrics
mkdir output/tables/time_to_discovery

mkdir output/simulation/descriptives
~/.local/pipx/venvs/asreview/bin/python scripts/data_describe.py generated_data/dataset_custom_priors.csv -o output/simulation/descriptives/data_stats_dataset_custom_priors.json
~/.local/pipx/venvs/asreview/bin/python scripts/data_describe.py generated_data/dataset_without_priors.csv -o output/simulation/descriptives/data_stats_dataset_without_priors.json

asreview simulate generated_data/dataset_custom_priors.csv -o output/simulation/state_files/sim_custom_priors.asreview --seed 165 --prior-idx 85 86 87 88 89 90 91 92 93 94 95 96 97 98 99 100 101 102 103 104 105 106 107 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125 126 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143 144 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161 162 163 164 165 166 167 168 169
asreview metrics output/simulation/state_files/sim_custom_priors.asreview -o output/simulation/metrics/metrics_sim_custom_priors.json --quiet

asreview simulate generated_data/dataset_without_priors.csv -o output/simulation/state_files/sim_minimal_priors.asreview --prior-seed 535 --seed 165
asreview metrics output/simulation/state_files/sim_minimal_priors.asreview -o output/simulation/metrics/metrics_sim_minimal_priors.json --quiet

# Generate plot and tables for dataset
~/.local/pipx/venvs/asreview/bin/python scripts/get_plot.py -s output/simulation/state_files/ -o output/figures/plot_recall_sim.png -l filename --hide-random
~/.local/pipx/venvs/asreview/bin/python scripts/merge_metrics.py -s output/simulation/metrics/ -o output/tables/metrics/metrics_sim.csv
~/.local/pipx/venvs/asreview/bin/python scripts/merge_tds.py -s output/simulation/metrics/ -o output/tables/time_to_discovery/tds_sim.csv
~/.local/pipx/venvs/asreview/bin/python scripts/merge_descriptives.py  -s output/simulation/descriptives/
```

So where are all the vectorisations done? I assume, I can repeat the `simulate` and `metrics` stage as often as I need it, without re-vectorizing?

**Reply by jteijema** [reply]
Hey @rkrug ,

The vectorization is done during the simulation, as is the testing of classification performance.

From what I'm reading, there's a few things I want to clarify:

ASReview isn't a tool for automatic classification of documents that are not labeled. It's a prioritization tool, that is, it will order the records based on the optimal reading order for finding all records, but we leave the final classification to the user. 

Technically you can use the api to get these rankings (from best to read first to worst to read first), and we always love experimentation, but there's some tricks.

You'd need to set the query strategy to max for example, other query strategies don't give the straightforward ranking (the optimal reading order might be different from the predicted ranking). 

The active learning cycle takes a label and retrains the classifier each time, so at what point do you take the ranking?

If you want to experiment with the api, I can recommend taking a look at the docs (https://asreview.readthedocs.io/en/stable/technical/example_api_asreview_file.html) and if you're using gen AI for help, definitely feed in some pages to seed the discussion. 


So just to recap, the simulation vectorizes and runs the classifier cycles all in one, which needs all labels. Automatically labeling partially labeled datasets isn't provided in any frontend or CLI options, as we want to help users classify, not classify ourselves. That doesn't mean it isn't available though,  but you'll need to use the python API.

Good luck!
^JT

**Comment by rkrug** [comment]
Thanks JT for the clarifications.

Final output: I would be fine with the ranking, especially if I could also get the underlying scores for the papers, so that I can use a certain cutoff.
At which point to take the ranking? I only have the priors - no further information, and I would like to have the rankings based on these. So I assume, it would be after the initial ranking?

The second part is the simulating of the priors (jackknifing or bootstrapping - whichever way you look at them) and to analyse how the threshold  used for the automatic classification (which I would do in R) impacts on the number of identified papers. This would be used to see if the priors need to be refined.

I guess that all this could probably done more easily with a custom implementation of embedding / vectorisation and then us a basic RAG implementation to identify the differences between each document and the labeled ones? But I would prefer to build this on the rather established foundation of ASReview instead of re-implementing this.

**Reply by jteijema** [reply]
Your plan can work. If you want to vary priors and avoid re-embedding each run, use the models directly rather than the simulation CLI. Concretely: take a feature extractor and a classifier, fit once on your corpus, cache the embeddings, then refit only the classifier for each bootstrap of priors.

When to take the ranking and scores: right after fitting the classifier on your current priors. Use the classifier’s predict_proba to get relevance scores and sort.

Fit the feature extractor once on the full corpus, transform to X, persist both fe and X. For each new prior set, reuse X, compute weights, refit the classifier, and rescore.

```python
from asreview.models.feature_extractors import Tfidf  # or SBERT
from asreview.models.classifiers import NaiveBayes    # or Logistic, SVM
from asreview.models.balancers import Balanced
from asreview import load_dataset
import pandas as pd
import numpy as np

# 1) Load and prepare a single dataframe with a binary column 'label_included'
asr_search = load_dataset("https://osf.io/download/zf95a/").get_df()
asr_inclusions = load_dataset("https://osf.io/download/j8stn/").get_df()
asr_inclusions["label_included"] = 1
asr_search["label_included"] = 0
df = pd.concat([asr_inclusions, asr_search], ignore_index=True)

# 2) Feature extraction once
fe = Tfidf()  # swap for SBERT() if you want semantic embeddings
X = fe.fit_transform(df)  # TextMerger handles title/abstract

# 3) Fit a classifier on your current priors
y = df["label_included"].values
ratio = (len(df) - y.sum()) / max(y.sum(), 1)  # simple neg/pos ratio
weights = Balanced(ratio).compute_sample_weight(y)

cls = NaiveBayes()
cls.fit(X, y, sample_weight=weights)

# 4) Get scores and ranking
scores = cls.predict_proba(X)[:, 1]  # relevance score
ranked = df.assign(score=scores).sort_values("score", ascending=False)

# Use 'score' cutoff in R as you like
```
